### PR TITLE
Add Amazon Bedrock Knowledge base as a RAG Engine (retriever)

### LINKED ISSUE #271: Feature: add Bedrock Knowledge Base as RAG Engine
Bedrock provides a unified document retrieve API https://docs.aws.amazon.com/bedrock/latest/APIReference/API_agent-runtime_Retrieve.html that can be integrated to the solution. This would provide out of the box support for all the Vector stores supported by Bedrock.

### PR DESCRIPTION
**Pull request by massi-ang**
*Issue #, if available:*
Resolves #271 

*Description of changes:*
With this PR we are adding Amazon Bedrock Knowledge Bases as additional RAG Engine (retrievers).

The KB must be already created and available in the same region used by Bedrock. The integration supports both semantic retrieval and hybrid retrieval via the KB retrieve API.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by hghandri** [comment]
Hi,

I'm really interested by this feature, do you have any ETA for integration on main branch ?

**Comment by joysl** [comment]
can I build the chatbot off of this branch?

**Comment by massi-ang** [comment]
> can I build the chatbot off of this branch?

Yes, you can now use this branch to deploy as it has been synched with `main`

**Comment by joysl** [comment]
@massi-ang Thank you so much, this worked for me

File: lib/shared/layers/python-sdk/python/genai_core/bedrock_kb/client.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            client = boto3.client(**config_data)
+
+            return client
+
+    raise genai_core.types.CommonError(
+        f"Could not find Amazon Bedrock KnowledgeBase ID {knowledge_base_id}"
+    )
+
+
+def get_kb_client_for_id(knowledge_base_id: str):
```
Comment (charles-marion) [inline lib/shared/layers/python-sdk/python/genai_core/bedrock_kb/client.py:49]: Not used and copy of the above.

File: lib/shared/layers/python-sdk/python/genai_core/workspaces.py [hunk 2]
```diff
@@ -188,7 +189,7 @@ def create_workspace_open_search(
         "format_version": 1,
         "name": workspace_name,
         "engine": "opensearch",
-        "status": "submitted",
+        "status": WorkspaceStatus.SUBMITTED,
```
Comment (charles-marion) [inline lib/shared/layers/python-sdk/python/genai_core/workspaces.py:192]: should be WorkspaceStatus.SUBMITTED.value, ?

File: lib/user-interface/react-app/public/aws-exports.json.eu-west-1 [hunk 3]
```diff
@@ -0,0 +1 @@
+{"aws_project_region":"eu-west-1","aws_cognito_region":"eu-west-1","aws_user_pools_id":"eu-west-1_xO3t84ZAv","aws_user_pools_web_client_id":"2j741c580dfp7ihp7aej8ct1nk","aws_cognito_identity_pool_id":"eu-west-1:fcd4f53c-d742-416d-8b1c-ebcfbaf50649","Auth":{"region":"eu-west-1","userPoolId":"eu-west-1_xO3t84ZAv","userPoolWebClientId":"2j741c580dfp7ihp7aej8ct1nk","identityPoolId":"eu-west-1:fcd4f53c-d742-416d-8b1c-ebcfbaf50649"},"aws_appsync_graphqlEndpoint":"https://45cdpihf6rcwxp5nq2m3k7l7zu.appsync-api.eu-west-1.amazonaws.com/graphql","aws_appsync_region":"eu-west-1","aws_appsync_authenticationType":"AMAZON_COGNITO_USER_POOLS","aws_appsync_apiKey":"da2-7k5rgwagwveqrebhqg7s7dbupq","Storage":{"AWSS3":{"bucket":"genaichatbotstack-chatbotapichatbucketsfilesbucke-1bhg45eovnhtw","region":"eu-west-1"}},"config":{"rag_enabled":true,"cross_encoders_enabled":true,"sagemaker_embeddings_enabled":true,"default_embeddings_model":"sagemaker::1024::intfloat/multilingual-e5-large","default_cross_encoder_model":"sagemaker::cross-encoder/ms-marco-MiniLM-L-12-v2"}}
```
Comment (charles-marion) [inline lib/user-interface/react-app/public/aws-exports.json.eu-west-1:1]: Should be removed / gitignored

File: lib/user-interface/react-app/public/aws-exports.json.us-east-1 [hunk 4]
```diff
@@ -0,0 +1 @@
+{"aws_project_region":"us-east-1","aws_cognito_region":"us-east-1","aws_user_pools_id":"us-east-1_jS3JLsdao","aws_user_pools_web_client_id":"4rssa2fqltnl056bft0mcgoc91","aws_cognito_identity_pool_id":"us-east-1:b7fc6357-61c1-4df3-bc8f-eb39496a90df","Auth":{"region":"us-east-1","userPoolId":"us-east-1_jS3JLsdao","userPoolWebClientId":"4rssa2fqltnl056bft0mcgoc91","identityPoolId":"us-east-1:b7fc6357-61c1-4df3-bc8f-eb39496a90df"},"aws_appsync_graphqlEndpoint":"https://kxetbl62kbhhvkyjfp52vwp5di.appsync-api.us-east-1.amazonaws.com/graphql","aws_appsync_region":"us-east-1","aws_appsync_authenticationType":"AMAZON_COGNITO_USER_POOLS","aws_appsync_apiKey":"da2-gzhezygcsnczpcloxnzohzsehe","Storage":{"AWSS3":{"bucket":"genaichatbotstack-chatbotapichatbucketsfilesbucket-liyvhjc0vsy2","region":"us-east-1"}},"config":{"rag_enabled":true,"cross_encoders_enabled":true,"sagemaker_embeddings_enabled":true,"default_embeddings_model":"sagemaker::1024::intfloat/multilingual-e5-large","default_cross_encoder_model":"sagemaker::cross-encoder/ms-marco-MiniLM-L-12-v2","privateWebsite":false}}
```
Comment (charles-marion) [inline lib/user-interface/react-app/public/aws-exports.json.us-east-1:1]: Should be removed / gitignored

File: lib/user-interface/react-app/src/pages/rag/create-workspace/create-workspace.tsx [hunk 5]
```diff
@@ -31,7 +32,7 @@ export default function CreateWorkspace() {
       const apiClient = new ApiClient(appContext);
       try {
         const result = await apiClient.ragEngines.getRagEngines();
-
+        console.log(result);
```
Comment (charles-marion) [inline lib/user-interface/react-app/src/pages/rag/create-workspace/create-workspace.tsx:35]: nit: debug code?

File: lib/user-interface/react-app/src/pages/rag/create-workspace/kb-form.tsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  >([]);
+
+  useEffect(() => {
+    if (!appContext) return;
+
+    (async () => {
+      const apiClient = new ApiClient(appContext);
+      try {
+        const result = await apiClient.bedrockKB.listKnowledgeBases();
+        console.log(result);
```
Comment (charles-marion) [inline lib/user-interface/react-app/src/pages/rag/create-workspace/kb-form.tsx:43]: debug code?

File: lib/user-interface/react-app/src/pages/rag/create-workspace/kb-form.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            }
+          />
+        </FormField>
+        <HybridSearchField
+          submitting={props.submitting}
+          errors={props.errors}
+          checked={props.data.hybridSearch}
+          onChange={props.onChange}
+        />
+        <MetadataFilter />
```
Comment (charles-marion) [inline lib/user-interface/react-app/src/pages/rag/create-workspace/kb-form.tsx:108]: Are these filter expressions used ? I did not find the callback.

**Comment by bigadsoleiman** [comment]
There are few conflicts that needs to be resolved first

File: bin/config copy.json [hunk 8]
```diff
@@ -0,0 +1,84 @@
+{
```
[the remark this answers is not in the document]
Comment (massi-ang) [reply bin/config copy.json]: fixed

File: lib/user-interface/react-app/src/pages/rag/create-workspace/kb-form.tsx (hunk 6 above)
Comment (massi-ang) [reply lib/user-interface/react-app/src/pages/rag/create-workspace/kb-form.tsx:43]: will remove

File: lib/shared/layers/python-sdk/python/genai_core/bedrock_kb/client.py (hunk 1 above)
Comment (massi-ang) [reply lib/shared/layers/python-sdk/python/genai_core/bedrock_kb/client.py:49]: It is not the same as the above, but you are right in pointing out that currently it is not used. I'll remove it

File: lib/chatbot-api/functions/api-handler/routes/bedrock_kb.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+from aws_lambda_powertools import Logger, Tracer
+from aws_lambda_powertools.event_handler.appsync import Router
+
+tracer = Tracer()
+router = Router()
+logger = Logger()
+
+
+class KendraDataSynchRequest(BaseModel):
+    workspaceId: str
```
[the remark this answers is not in the document]
Comment (massi-ang) [reply lib/chatbot-api/functions/api-handler/routes/bedrock_kb.py:13]: I would prefer this is addressed systematically for all APIs.

File: .codecatalyst/workflows/Workflow_81ec.yaml [hunk 10]
```diff
@@ -0,0 +1,29 @@
+Name: Workflow_81ec
```
[the remark this answers is not in the document]
Comment (massi-ang) [reply .codecatalyst/workflows/Workflow_81ec.yaml:1]: will remove

**Review by charles-marion — APPROVED** [review]

**Pull request merged**